### ISSUE TITLE
About the attention

### ISSUE THREAD

**Issue by LCLL** [Rayhane-mamah/Tacotron-2#1]
Hi, Rayhane-mamah,
      Thanks for sharing the code. I have got a acceptable result in tacotron before, but sometimes the tacotron model produces repeated words or misses some words. In tacotron2 paper, the authors say that they use a different attention model to handle these problems. So I replace the attention model in tacotron with relative code used in tacotron2. However, I can't get promising result anymore. Even that the model can not produce correct voices given by the text. Can you give me some suggestions about that ?

**Comment by Rayhane-mamah** [comment]
Hi @LCLL, thank you for reaching out.

To be able to give you a correct answer, I have to ask the following questions:
- Did you try using the attention mechanism of our Tacotron 2 implementation?
- Can you provide some alignment plots of your model?
- Did you only modify the attention mechanism and keep the original Tacotron architecture?

**Comment by LCLL** [comment]
Hi @Rayhane-mamah  Thank you for your quick reply:) I think I made mistakes on the attention mechanism replacement. I used the tacotron code provided in https://github.com/keithito/tacotron.

I just copied the relative attention part (_location_based_score, LocationBasedAttention) in attention_wrapper.py into /tensorflow/contrib/seq2seq/python/ops/attention_wrapper.py. But I think it's not enough because I found that the AttentionWrapper is also changed compared to the tensorflow original version. Is there a convenient way to reach that if I just want to replace the attention part? I want to check if the located based attention could handle the missing/repeated words problems.

**Comment by Rayhane-mamah** [comment]
@LCLL, about the changes made to attention_wrapper.py, rnn_wrappers.py, custom_decoder.py, dynamic_decoder.py or helpers.py, they were mainly made to fit with the Tacotron 2 architecture described in [Natural TTS synthesis by conditioning Wavenet on MEL spectogram predictions](https://arxiv.org/pdf/1712.05884.pdf)

The (_location_based_score, LocationBasedAttention) I added to the attention_wrapper.py should work fine without the need of changing anything else in [keithito's tacotron implementation](https://github.com/keithito/tacotron).

For that matter, it would really help us knowing the cause of the issue if you could provide the alignment plots given by your training script. It's usually the best practice to catch the problems related to attention mechanisms.

Also knowing knowing how many training steps you did, which dataset you used can be helpful too.

**Comment by LCLL** [comment]
OK. I removed the alignments image because I think the result is not right. I have trained for 3W+ steps and found the big difference between the original tacotron version and the modified version. In the original version, it took about 8k steps to get a reasonable alignments and I can use the model to get meaningful voices by providing arbitrary texts. However, I can't get the same result even in more than 3W+ steps. The loss is decreasing but the alignments are still like the one in the beginning of the training.
I re-run the code and will post the alignments result ASAP:)
Thanks.

**Comment by Rayhane-mamah** [comment]
We will be looking at the issue too and we'll keep you informed.

**Comment by imdatsolak** [comment]
@Rayhane-mamah, I'm currently starting work on using a WaveNet based Vocoder to connect to Tacotron-2. Any pointers to your Tacotron-2 implementation to start with? I have a wavenet_vocoder implementation that could use Mel- or WAV-conditioning to generate audio. I would like to put a Mel as input to WaveNet and have then a final audio as output... Any recommendation would be appreciated. At the end, I want to make a parallel-wavenet out of it and then put back on GitHub.

**Comment by LCLL** [comment]
I compared LuongAttention(BahdanauAttention is similar to LuongAttention) with LocationBasedAttention.
LuongAttention:
Total Loss:
![luong_loss_total](https://user-images.githubusercontent.com/3695454/35422968-485e49f2-0286-11e8-9c5d-aaa7631402b8.jpg)
Mel Loss:
![luong_loss_mel](https://user-images.githubusercontent.com/3695454/35422977-5ab2fb16-0286-11e8-9f71-92969c394ff8.jpg)
Linear Loss:
![luong_loss_linear](https://user-images.githubusercontent.com/3695454/35422980-5ea86922-0286-11e8-9950-63acf1e2a00a.jpg)

LocationBasedAttention:
Total Loss:
![location_loss_total](https://user-images.githubusercontent.com/3695454/35422986-6986b24a-0286-11e8-9ab5-92dc883ccf7d.jpg)
Mel Loss:
![location_loss_mel](https://user-images.githubusercontent.com/3695454/35422992-715eac7a-0286-11e8-99ea-afbf531f590e.jpg)
Linear Loss:
![location_loss_linear](https://user-images.githubusercontent.com/3695454/35422996-77156d2a-0286-11e8-8d0a-b709c5abf11d.jpg)

Alignments:
6k steps:
LuongAttention:
![step-6000-align](https://user-images.githubusercontent.com/3695454/35423004-8c6e49e4-0286-11e8-9313-f7f585983300.png)
LocationBasedAttention:
![step-6000-align](https://user-images.githubusercontent.com/3695454/35423012-960dd8f2-0286-11e8-9a75-cab1b4c9e137.png)

1w steps:
LuongAttention:
![step-10000-align](https://user-images.githubusercontent.com/3695454/35423028-a8e9bbc6-0286-11e8-9757-de653b734d13.png)
LocationBasedAttention:
![step-10000-align](https://user-images.githubusercontent.com/3695454/35423036-b0e31f20-0286-11e8-86e3-076d2f193a49.png)

1w5 steps:
LuongAttention:
![step-15000-align](https://user-images.githubusercontent.com/3695454/35423044-bc972442-0286-11e8-9cb0-e09ac36f50dd.png)
LocationBasedAttention:
![step-15000-align](https://user-images.githubusercontent.com/3695454/35423053-c3fca518-0286-11e8-89f3-f389bb988a79.png)

2w7 steps:
LuongAttention:
![step-27000-align](https://user-images.githubusercontent.com/3695454/35423075-d796165e-0286-11e8-9d21-54414f0476bb.png)
LocationBasedAttention:
![step-27000-align](https://user-images.githubusercontent.com/3695454/35423085-e168df22-0286-11e8-9cfb-c5eaa6f4f5ad.png)

It's strange that the total loss of LocationBasedAttention is increasing suddenly around 4.8k step, 1.3w step, 2.6w step, caused by the increasing of mel loss.

**Comment by LCLL** [comment]
What's more, I used the minimal loss checkpoint (around 2.5w step) of both models to eval the performance, and found that the LocationBasedAttention model missed words in many test cases, in contrast, the LuongAttention model has a much better performance. I wonder if I have to change some parameters in tacotron to improve the LocationBasedAttention model.

**Comment by Rayhane-mamah** [comment]
@LCLL, from the plots it's pretty clear that for some reason, the model when using the Location base attention have some loss peaks, and looses that's probably due to the model loosing its attention progress.

After revisiting our work implementing Tacotron-2 and the original Tacotron architecture, there are hypotheses that come up and that we are investigating:

1- Tacotron-2 concatenates the context vector and the prenet output before feeding them to the RNN cells of the decoder, which is missing the original Tacotron implementation, the paper writers also specified that without the prenet, attention cannot be learned, I wonder if they wanted to insist on the concatenation part too.
2- Other than that point, everything is almost the same in the two architectures (if we ignore the nature of the cells in use and the number of units) which suggests that we made a mistake somewhere inside our attention mechanism, our first guess would be the concatenation of the decoder predicted frame with the context vector before entering the prenet (this was not explicitly described in the paper but we still kept it). Our second guess would be to revisit the attention implementation and verify its concordance with the paper (which we did several times).
3- If the two first hypotheses are wrong, then what's left to check is if additional training steps or toying with the hyper parameters can improve the model. It is important to keep in mind that the two architectures use different loss functions. (which may explain why we never had the explosion phenomena in our tests)

With this said, we are really excited to test what would work best and have a better understanding of the issue, but since our computation power is currently taken for other projects, we won't be able to search deeper and efficiently before the end of February. Please inform us however of any progress you make.
